File: products/menu/scripts/explore-idealista.ts
import { chromium } from "@playwright/test";
import path from "path";
import fs from "fs";

const EMAIL = "eduardoferdcarvalho+agency@gmail.com";
const PASSWORD = "Teste1234_@";

const SCREENSHOTS = path.join(import.meta.dir, "idealista-screenshots");
fs.mkdirSync(SCREENSHOTS, { recursive: true });

async function shot(page: any, name: string) {
  const p = path.join(SCREENSHOTS, `${name}.png`);
  await page.screenshot({ path: p, fullPage: false });
  console.log(`📸 ${name}`);
}

const browser = await chromium.launch({
  headless: true,
  executablePath: "C:\\Program Files\\Google\\Chrome\\Application\\chrome.exe",
});
const ctx = await browser.newContext({
  userAgent:
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36",
  viewport: { width: 1280, height: 900 },
  locale: "pt-PT",
});
const page = await ctx.newPage();

// ── 1. Homepage ──────────────────────────────────────────────────────────────
console.log("\n── 1. Homepage");
await page.goto("https://www.idealista.pt", { waitUntil: "domcontentloaded" });
await page.waitForTimeout(2000);
await shot(page, "01-homepage");

// Accept cookies if present
const cookieBtn = page.locator(
  "button:has-text('Aceitar'), button:has-text('Accept'), #didomi-notice-agree-button"
);
if (await cookieBtn.first().isVisible({ timeout: 3000 }).catch(() => false)) {
  await cookieBtn.first().click();
  console.log("  ✓ Cookies accepted");
  await page.waitForTimeout(1000);
}

// ── 2. Find login link ────────────────────────────────────────────────────────
console.log("\n── 2. Looking for login");
await shot(page, "02-before-login");

const loginLink = page.locator(
  "a[href*='login'], a[href*='acesso'], button:has-text('Entrar'), a:has-text('Entrar'), a:has-text('Iniciar sessão')"
);
const loginVisible = await loginLink
  .first()
  .isVisible({ timeout: 4000 })
  .catch(() => false);

if (loginVisible) {
  console.log("  ✓ Login link found, clicking…");
  await loginLink.first().click();
  await page.waitForTimeout(2000);
  await shot(page, "03-login-page");
} else {
  console.log("  ⚠ No login link visible, navigating directly…");
  await page.goto("https://www.idealista.pt/areas/login/", {
    waitUntil: "domcontentloaded",
  });
  await page.waitForTimeout(2000);
  await shot(page, "03-login-page-direct");
}

// ── 3. Fill credentials ───────────────────────────────────────────────────────
console.log("\n── 3. Filling credentials");
const emailField = page.locator(
  'input[type="email"], input[name="email"], input[id*="email"], input[name="username"]'
);
const passField = page.locator('input[type="password"]');

if (await emailField.first().isVisible({ timeout: 4000 }).catch(() => false)) {
  await emailField.first().fill(EMAIL);
  await passField.first().fill(PASSWORD);
  await shot(page, "04-credentials-filled");

  const submitBtn = page.locator(
    'button[type="submit"], input[type="submit"], button:has-text("Entrar"), button:has-text("Iniciar")'
  );
  await submitBtn.first().click();
  await page.waitForTimeout(3000);
  await shot(page, "05-after-login");
  console.log(`  Current URL: ${page.url()}`);
} else {
  console.log("  ⚠ No email field found");
  await shot(page, "04-no-email-field");
}

// ── 4. Explore post-login area ───────────────────────────────────────────────
console.log("\n── 4. Post-login exploration");
const currentUrl = page.url();
console.log(`  URL: ${currentUrl}`);

// Look for "anunciar" / "publicar" / "add listing" links
const publishLinks = page.locator(
  "a[href*='anunciar'], a[href*='publicar'], a[href*='anunci'], a:has-text('Anunciar'), a:has-text('Publicar'), a:has-text('Novo anúncio')"
);
const publishCount = await publishLinks.count();
console.log(`  Found ${publishCount} publish-related links`);

if (publishCount > 0) {
  for (let i = 0; i < Math.min(publishCount, 3); i++) {
    const href = await publishLinks.nth(i).getAttribute("href");
    const text = await publishLinks.nth(i).innerText().catch(() => "");
    console.log(`    [${i}] "${text.trim()}" → ${href}`);
  }

  // Click the first one
  await publishLinks.first().click();
  await page.waitForTimeout(2000);
  await shot(page, "06-publish-flow");
  console.log(`  Publish URL: ${page.url()}`);

  // One more level deep
  await shot(page, "07-publish-detail");
}

// ── 5. Check for agency/pro area ─────────────────────────────────────────────
console.log("\n── 5. Agency/pro area");
const proLinks = page.locator(
  "a[href*='profissional'], a[href*='agencia'], a[href*='empresa'], a:has-text('Profissional'), a:has-text('Agência')"
);
const proCount = await proLinks.count();
if (proCount > 0) {
  for (let i = 0; i < Math.min(proCount, 3); i++) {
    const href = await proLinks.nth(i).getAttribute("href");
    const text = await proLinks.nth(i).innerText().catch(() => "");
    console.log(`    [${i}] "${text.trim()}" → ${href}`);
  }
}

// ── 6. Network requests snapshot ─────────────────────────────────────────────
console.log("\n── 6. Checking for API endpoints during page load…");
const apiCalls: string[] = [];
page.on("request", (req) => {
  const url = req.url();
  if (url.includes("/api/") || url.includes(".json") || url.includes("graphql")) {
    apiCalls.push(`${req.method()} ${url}`);
  }
});

// Navigate to dashboard/account area
await page.goto("https://www.idealista.pt/areas/minha-area/", {
  waitUntil: "domcontentloaded",
});
await page.waitForTimeout(2000);
await shot(page, "08-my-area");
console.log(`  My-area URL: ${page.url()}`);

if (apiCalls.length > 0) {
  console.log("\n  API calls intercepted:");
  apiCalls.forEach((c) => console.log(`    ${c}`));
}

// Final state
await shot(page, "09-final");
console.log(`\n✅ Done. Screenshots saved to: ${SCREENSHOTS}`);

await browser.close();

File: products/menu/scripts/explore-idealista.py
import sys
sys.stdout.reconfigure(encoding="utf-8")
import undetected_chromedriver as uc
import time
import os
import json

EMAIL = "eduardoferdcarvalho+agency@gmail.com"
PASSWORD = "Teste1234_@"
SCREENSHOTS = os.path.join(os.path.dirname(__file__), "idealista-screenshots")
os.makedirs(SCREENSHOTS, exist_ok=True)

def shot(driver, name):
    p = os.path.join(SCREENSHOTS, f"{name}.png")
    driver.save_screenshot(p)
    print(f"[screenshot] {name}")

def body_text(driver, chars=300):
    try:
        return driver.find_element("tag name", "body").text[:chars]
    except:
        return driver.page_source[:chars]

print("Launching undetected Chrome")
options = uc.ChromeOptions()
options.add_argument("--window-size=1280,900")
options.add_argument("--lang=pt-PT")

driver = uc.Chrome(
    options=options,
    headless=2,
    use_subprocess=True,
    version_main=148,
    browser_executable_path=r"C:\Program Files\Google\Chrome\Application\chrome.exe",
)

try:
    #  1. Homepage 
    print("\n 1. Homepage")
    driver.get("https://www.idealista.pt")
    time.sleep(4)
    shot(driver, "uc01-homepage")
    print(f"  Title: {driver.title}")
    print(f"  URL:   {driver.current_url}")
    print(f"  Body:  {body_text(driver, 150)}")

    # Accept cookies
    try:
        btn = driver.find_element("id", "didomi-notice-agree-button")
        btn.click()
        print("  OK Cookies accepted")
        time.sleep(1)
    except:
        pass

    shot(driver, "uc02-homepage-clean")

    #  2. Login page 
    print("\n 2. Login")
    driver.get("https://www.idealista.pt/areas/login/")
    time.sleep(4)
    shot(driver, "uc03-login")
    print(f"  URL: {driver.current_url}")
    print(f"  Body: {body_text(driver, 200)}")

    inputs = driver.find_elements("tag name", "input")
    print(f"  Inputs: {[{'type': i.get_attribute('type'), 'name': i.get_attribute('name'), 'id': i.get_attribute('id')} for i in inputs]}")

    email_field = None
    for i in inputs:
        if i.get_attribute("type") in ("email", "text") or i.get_attribute("name") in ("email", "username"):
            email_field = i
            break

    if email_field:
        #  3. Fill & submit 
        print("\n 3. Filling credentials")
        email_field.clear()
        email_field.send_keys(EMAIL)
        pass_field = driver.find_element("css selector", 'input[type="password"]')
        pass_field.send_keys(PASSWORD)
        shot(driver, "uc04-filled")
        submit = driver.find_element("css selector", 'button[type="submit"], input[type="submit"]')
        submit.click()
        time.sleep(4)
        shot(driver, "uc05-after-login")
        print(f"  Post-login URL: {driver.current_url}")

        #  4. My area 
        print("\n 4. My area")
        driver.get("https://www.idealista.pt/areas/minha-area/")
        time.sleep(3)
        shot(driver, "uc06-my-area")
        print(f"  URL: {driver.current_url}")
        print(f"  Body:\n{body_text(driver, 600)}")

        links = driver.find_elements("css selector", "a[href]")
        print("\n  Links on page:")
        seen = set()
        for l in links[:40]:
            href = l.get_attribute("href") or ""
            text = l.text.strip()[:60]
            if text and href and href not in seen and "javascript" not in href:
                seen.add(href)
                print(f"    \"{text}\"  {href}")

        #  5. Create listing 
        print("\n 5. Anunciar (publish listing)")
        driver.get("https://www.idealista.pt/areas/anunciar/")
        time.sleep(3)
        shot(driver, "uc07-anunciar")
        print(f"  URL: {driver.current_url}")
        print(f"  Body:\n{body_text(driver, 800)}")

        form_fields = driver.find_elements("css selector", "input, select, textarea")
        if form_fields:
            print("  Form fields:")
            for f in form_fields:
                print(f"    <{f.tag_name} type='{f.get_attribute('type')}' name='{f.get_attribute('name')}' id='{f.get_attribute('id')}' placeholder='{f.get_attribute('placeholder')}'>")

        shot(driver, "uc08-anunciar-detail")

        # Also try the pro/agency area
        print("\n 6. Pro/agency area")
        for url in [
            "https://www.idealista.pt/areas/minha-area/anuncios/",
            "https://www.idealista.pt/areas/minha-area/perfil/",
        ]:
            driver.get(url)
            time.sleep(2)
            slug = url.rstrip("/").split("/")[-1]
            shot(driver, f"uc09-{slug}")
            print(f"  {url}  {driver.current_url}")
            print(f"  {body_text(driver, 300)}\n")

    else:
        print("  WARN DataDome still blocking  no login form found")
        print(f"  Page source snippet:\n{driver.page_source[:800]}")

finally:
    driver.quit()
    print(f"\nDONE Done. Screenshots: {SCREENSHOTS}")
